File: apps/web/src/simulation/hierarchy.ts
import { initialOutput } from "./conditional";
import {
    edgeId,
    type Block,
    type ModelDocument,
    type Connection,
    type Port,
} from "./model";

export function descendants(
    blocks: Block[],
    selected: ReadonlySet<string>,
): Set<string> {
    const result = new Set(selected),
        children = new Map<string, string[]>();
    for (const b of blocks)
        if (b.parent)
            children.set(b.parent, [...(children.get(b.parent) ?? []), b.id]);
    const queue = [...selected];
    for (let i = 0; i < queue.length; i++)
        for (const id of children.get(queue[i]!) ?? [])
            if (!result.has(id)) {
                result.add(id);
                queue.push(id);
            }
    return result;
}
export function validateHierarchy(blocks: Block[]) {
    const byId = new Map(blocks.map((b) => [b.id, b]));
    const groups = new Map<string, Block[]>();
    for (const b of blocks) {
        let parent = b.parent;
        const seen = new Set([b.id]);
        while (parent) {
            const p = byId.get(parent);
            if (
                !p ||
                p.kind.type !== "subsystem" ||
                seen.has(parent) ||
                seen.size > 32
            )
                throw new Error("父系统无效、存在循环或层级超过 32 层。");
            seen.add(parent);
            parent = p.parent;
        }
        if (b.kind.type === "inport" || b.kind.type === "outport") {
            if (b.parent && b.kind.type === "inport" && b.kind.data)
                throw new Error(
                    "子系统内部 Inport 从父系统接收信号，不能绑定独立数据。",
                );
            const key = JSON.stringify([b.parent, b.kind.type]);
            groups.set(key, [...(groups.get(key) ?? []), b]);
        }
    }
    for (const parent of [
        undefined,
        ...blocks.filter((b) => b.kind.type === "subsystem").map((b) => b.id),
    ])
        for (const type of ["inport", "outport"] as const) {
            const ports = groups.get(JSON.stringify([parent, type])) ?? [];
            const numbers = ports
                .map((b) => (b.kind as { port: number }).port)
                .sort((a, b) => a - b);
            const p = parent ? byId.get(parent) : undefined;
            const expected =
                p?.kind.type === "subsystem"
                    ? type === "inport"
                        ? p.kind.inputs
                        : p.kind.outputs
                    : ports.length;
            if (
                numbers.length > 64 ||
                numbers.length !== expected ||
                numbers.some((n, i) => n !== i + 1)
            )
                throw new Error(
                    "Inport / Outport 编号必须连续，并与子系统端口数量一致。",
                );
        }
}
export function renumberBoundaries(doc: ModelDocument) {
    const mappings = new Map<string, Map<string, string>>();
    for (const parent of [
        undefined,
        ...doc.model.blocks
            .filter((b) => b.kind.type === "subsystem")
            .map((b) => b.id),
    ])
        for (const type of ["inport", "outport"] as const) {
            const ports = doc.model.blocks
                .filter((b) => b.parent === parent && b.kind.type === type)
                .sort(
                    (a, b) =>
                        (a.kind as { port: number }).port -
                        (b.kind as { port: number }).port,
                );
            const map = mappings.get(parent ?? "") ?? new Map<string, string>();
            mappings.set(parent ?? "", map);
            const owner = doc.model.blocks.find((b) => b.id === parent);
            const execution =
                owner?.kind.type === "subsystem"
                    ? owner.kind.execution
                    : undefined;
            const policies =
                type === "outport" && execution
                    ? ports.map((b) =>
                          b.kind.type === "outport"
                              ? (execution.outputs[b.kind.port - 1] ??
                                initialOutput())
                              : initialOutput(),
                      )
                    : undefined;
            ports.forEach((b, i) => {
                if (b.kind.type !== "inport" && b.kind.type !== "outport")
                    return;
                map.set(
                    `${type === "inport" ? "in" : "out"}${b.kind.port}`,
                    `${type === "inport" ? "in" : "out"}${i + 1}`,
                );
                b.kind.port = i + 1;
            });
            const p = doc.model.blocks.find((b) => b.id === parent);
            if (p?.kind.type === "subsystem") {
                p.kind[type === "inport" ? "inputs" : "outputs"] = ports.length;
                if (policies && p.kind.execution)
                    p.kind.execution.outputs = policies;
            }
        }
    doc.model.connections = doc.model.connections.filter((edge) => {
        for (const p of [edge.from, edge.to]) {
            const map = mappings.get(p.block);
            if (map && p.port !== "enable" && p.port !== "trigger") {
                const next = map.get(p.port);
                if (!next) return false;
                p.port = next;
            }
        }
        return true;
    });
}
function upgrade(doc: ModelDocument) {
    doc.schemaVersion = doc.model.schemaVersion = Math.max(
        doc.schemaVersion,
        doc.model.schemaVersion,
        7,
    ) as ModelDocument["schemaVersion"];
}
const key = (p: Port) => JSON.stringify(p);
const wire = (from: Port, to: Port): Connection => ({
    from: { ...from },
    to: { ...to },
});

export function groupBlocks(
    doc: ModelDocument,
    selected: ReadonlySet<string>,
    idFactory: () => string,
): { document: ModelDocument; id: string } {
    const peers = doc.model.blocks.filter((b) => selected.has(b.id));
    if (
        !peers.length ||
        peers.some(
            (b) =>
                b.parent !== peers[0]!.parent ||
                b.kind.type === "inport" ||
                b.kind.type === "outport",
        )
    )
        throw new Error("请选择同一系统中的计算方块，保留现有边界端口。");
    const next = structuredClone(doc);
    upgrade(next);
    const id = idFactory(),
        parent = peers[0]!.parent;
    const incoming = new Map<string, { source: Port; port: Block }>(),
        outgoing = new Map<string, { source: Port; port: Block }>();
    const minX = Math.min(...peers.map((b) => b.position?.x ?? 0)),
        minY = Math.min(...peers.map((b) => b.position?.y ?? 0)),
        maxX = Math.max(...peers.map((b) => b.position?.x ?? 0));
    const replacements: Connection[] = [];
    for (const e of next.model.connections) {
        const a = selected.has(e.from.block),
            b = selected.has(e.to.block);
        if (a === b) {
            replacements.push(e);
            continue;
        }
        if (b) {
            let found = incoming.get(key(e.from));
            if (!found) {
                const port: Block = {
                    id: idFactory(),
                    parent: id,
                    position: { x: 30, y: 100 + incoming.size * 130 },
                    kind: { type: "inport", port: incoming.size + 1 },
                };
                found = { source: e.from, port };
                incoming.set(key(e.from), found);
                replacements.push(
                    wire(e.from, { block: id, port: `in${incoming.size}` }),
                );
            }
            replacements.push(
                wire({ block: found.port.id, port: "out" }, e.to),
            );
        } else {
            let found = outgoing.get(key(e.from));
            if (!found) {
                const port: Block = {
                    id: idFactory(),
                    parent: id,
                    position: {
                        x: 460 + maxX - minX,
                        y: 100 + outgoing.size * 130,
                    },
                    kind: { type: "outport", port: outgoing.size + 1 },
                };
                found = { source: e.from, port };
                outgoing.set(key(e.from), found);
                replacements.push(wire(e.from, { block: port.id, port: "in" }));
            }
            replacements.push(
                wire(
                    {
                        block: id,
                        port: `out${(found.port.kind as { port: number }).port}`,
                    },
                    e.to,
                ),
            );
        }
    }
    if (incoming.size > 64 || outgoing.size > 64)
        throw new Error("一个子系统最多支持 64 个输入和 64 个输出。");
    for (const b of next.model.blocks)
        if (selected.has(b.id)) {
            b.parent = id;
            b.position = {
                x: (b.position?.x ?? 0) - minX + 220,
                y: (b.position?.y ?? 0) - minY + 100,
            };
        }
    next.model.blocks.push(
        {
            id,
            ...(parent ? { parent } : {}),
            position: { x: minX, y: minY },
            kind: {
                type: "subsystem",
                inputs: incoming.size,
                outputs: outgoing.size,
            },
        },
        ...[...incoming.values(), ...outgoing.values()].map((e) => e.port),
    );
    next.editor.labels[id] = "Subsystem";
    for (const { port } of [...incoming.values(), ...outgoing.values()])
        next.editor.labels[port.id] =
            `${port.kind.type === "inport" ? "In" : "Out"}${(port.kind as { port: number }).port}`;
    next.model.connections = replacements;
    cleanBends(next);
    for (const edge of next.model.connections)
        if (selected.has(edge.from.block) || selected.has(edge.to.block))
            delete next.editor.bends[edgeId(edge)];
    validateHierarchy(next.model.blocks);
    return { document: next, id };
}
export function ungroupBlock(doc: ModelDocument, id: string): ModelDocument {
    const next = structuredClone(doc),
        sub = next.model.blocks.find((b) => b.id === id);
    if (sub?.kind.type !== "subsystem") throw new Error("请选择一个子系统。");
    if (sub.kind.execution)
        throw new Error(
            "条件子系统保留执行边界；如需展开，请先在检查器中改为普通子系统。",
        );
    const boundaries = next.model.blocks.filter(
        (b) =>
            b.parent === id &&
            (b.kind.type === "inport" || b.kind.type === "outport"),
    );
    const removed = new Set([id, ...boundaries.map((b) => b.id)]),
        drivers = new Map(
            next.model.connections.map((e) => [key(e.to), e.from]),
        );
    const aliases = new Map<string, Port | undefined>();
    for (const b of boundaries) {
        if (b.kind.type === "inport")
            aliases.set(
                key({ block: b.id, port: "out" }),
                drivers.get(key({ block: id, port: `in${b.kind.port}` })),
            );
        else if (b.kind.type === "outport")
            aliases.set(
                key({ block: id, port: `out${b.kind.port}` }),
                drivers.get(key({ block: b.id, port: "in" })),
            );
    }
    const resolve = (source: Port): Port | undefined => {
        let p: Port | undefined = source;
        const seen = new Set<string>();
        while (p && aliases.has(key(p))) {
            if (seen.has(key(p)))
                throw new Error("边界端口存在循环，无法展开。");
            seen.add(key(p));
            p = aliases.get(key(p));
        }
        return p;
    };
    next.model.connections = next.model.connections.flatMap((e) => {
        if (removed.has(e.to.block)) return [];
        const source = resolve(e.from);
        return source ? [wire(source, e.to)] : [];
    });
    next.model.blocks = next.model.blocks.filter((b) => !removed.has(b.id));
    const moved = new Set(
        next.model.blocks.filter((b) => b.parent === id).map((b) => b.id),
    );
    for (const b of next.model.blocks)
        if (b.parent === id) {
            if (sub.parent) b.parent = sub.parent;
            else delete b.parent;
            b.position = {
                x: (b.position?.x ?? 0) + (sub.position?.x ?? 0) - 220,
                y: (b.position?.y ?? 0) + (sub.position?.y ?? 0) - 100,
            };
        }
    for (const removedId of removed) {
        delete next.editor.labels[removedId];
        if (next.model.sampleTimes) delete next.model.sampleTimes[removedId];
    }
    cleanBends(next);
    for (const edge of next.model.connections)
        if (moved.has(edge.from.block) || moved.has(edge.to.block))
            delete next.editor.bends[edgeId(edge)];
    validateHierarchy(next.model.blocks);
    return next;
}
export function cleanBends(doc: ModelDocument) {
    const kept = new Set(doc.model.connections.map(edgeId));
    for (const id of Object.keys(doc.editor.bends))
        if (!kept.has(id)) delete doc.editor.bends[id];
}

File: simulation/crates/openmat-sim/src/authoring/hierarchy.rs
use crate::{
    ModelError,
    model::{Block, BlockKind, Connection, Model, Port},
};
use std::collections::{BTreeMap, BTreeSet};

type Key = (String, String);
fn key(p: &Port) -> Key {
    (p.block.clone(), p.port.clone())
}
fn fail(b: &str, message: &str) -> ModelError {
    ModelError::new("hierarchy", message).at(b, None)
}

fn collect_domains(
    model: &Model,
    blocks: &Blocks<'_>,
    boundaries: &Boundaries<'_>,
    drivers: &BTreeMap<Key, Port>,
) -> Result<Vec<crate::conditional::DomainSpec>, ModelError> {
    let mut domains = Vec::new();
    for group in &model.blocks {
        let BlockKind::Subsystem {
            execution: Some(execution),
            outputs,
            ..
        } = &group.kind
        else {
            continue;
        };
        let mut members = Vec::new();
        let mut scopes = Vec::new();
        for b in &model.blocks {
            let mut parent = b.parent.as_deref();
            while let Some(p) = parent {
                if p == group.id {
                    if matches!(
                        b.kind,
                        BlockKind::Subsystem {
                            execution: Some(_),
                            ..
                        }
                    ) {
                        return Err(fail(
                            &b.id,
                            "nested conditional domains are not supported in conditional-v1",
                        ));
                    }
                    if !matches!(
                        b.kind,
                        BlockKind::Subsystem { .. }
                            | BlockKind::Inport { .. }
                            | BlockKind::Outport { .. }
                    ) {
                        members.push(b.id.clone());
                        if matches!(b.kind, BlockKind::Scope) {
                            scopes.push(b.id.clone());
                        }
                    }
                    break;
                }
                parent = blocks[p].parent.as_deref();
            }
        }
        let outports: Vec<_> = (1..=*outputs)
            .map(|p| boundaries[&(Some(group.id.as_str()), false)][&p].id.clone())
            .collect();
        members.extend(outports.iter().cloned());
        if members.is_empty() || domains.len() >= 64 {
            return Err(fail(
                &group.id,
                "conditional domains need a computational body; at most 64 domains are supported",
            ));
        }
        domains.push(crate::conditional::DomainSpec {
            id: group.id.clone(),
            execution: execution.clone(),
            control: drivers[&(group.id.clone(), execution.control_port().into())].clone(),
            members,
            outports,
            scopes,
        });
    }
    Ok(domains)
}

pub(crate) fn flatten(
    model: &Model,
) -> Result<(Model, Vec<crate::conditional::DomainSpec>), ModelError> {
    let blocks: BTreeMap<_, _> = model.blocks.iter().map(|b| (b.id.as_str(), b)).collect();
    let boundaries = validate_hierarchy(model, &blocks)?;
    let drivers = validate_connections(model, &blocks)?;
    let mut domains = collect_domains(model, &blocks, &boundaries, &drivers)?;
    let conditional_outputs: BTreeSet<_> = domains
        .iter()
        .flat_map(|d| d.outports.iter().cloned())
        .collect();
    let conditional_scopes: BTreeSet<_> = domains
        .iter()
        .flat_map(|d| d.scopes.iter().cloned())
        .collect();
    let aliases = boundary_aliases(model, &boundaries, &drivers, &conditional_outputs);
    let mut cache: BTreeMap<Key, Port> = BTreeMap::new();
    let mut resolve = |source: &Port| resolve_alias(source, &aliases, &mut cache);
    // Validate every alias even if no observable consumes it.
    for (block, port) in aliases.keys() {
        resolve(&Port {
            block: block.clone(),
            port: port.clone(),
        })?;
    }
    for domain in &mut domains {
        domain.control = resolve(&domain.control)?;
    }
    let mut flat = model.clone();
    flat.blocks.retain(|b| {
        !(matches!(b.kind, BlockKind::Subsystem { .. })
            || b.parent.is_some()
                && matches!(b.kind, BlockKind::Inport { .. } | BlockKind::Outport { .. })
                && !conditional_outputs.contains(&b.id))
    });
    let retained: BTreeSet<_> = flat.blocks.iter().map(|b| b.id.clone()).collect();
    flat.connections = model
        .connections
        .iter()
        .filter(|e| retained.contains(&e.to.block))
        .map(|e| {
            Ok(Connection {
                from: resolve(&e.from)?,
                to: e.to.clone(),
            })
        })
        .collect::<Result<_, ModelError>>()?;
    flat.sample_times.retain(|id, _| retained.contains(id));
    for b in &mut flat.blocks {
        b.parent = None;
        if conditional_scopes.contains(&b.id) {
            b.kind = BlockKind::Outport { port: 0 };
        } else if matches!(b.kind, BlockKind::Outport { .. })
            && !conditional_outputs.contains(&b.id)
        {
            b.kind = BlockKind::Scope;
        }
    }
    Ok((flat, domains))
}

fn boundary_aliases(
    model: &Model,
    boundaries: &Boundaries<'_>,
    drivers: &BTreeMap<Key, Port>,
    conditional_outputs: &BTreeSet<String>,
) -> BTreeMap<Key, Port> {
    let mut aliases = BTreeMap::new();
    for block in &model.blocks {
        match &block.kind {
            BlockKind::Inport { port, .. } if block.parent.is_some() => {
                aliases.insert(
                    (block.id.clone(), "out".into()),
                    drivers[&(block.parent.clone().expect("parent"), format!("in{port}"))].clone(),
                );
            }
            BlockKind::Subsystem { outputs, .. } => {
                for port in 1..=*outputs {
                    let child = boundaries[&(Some(block.id.as_str()), false)][&port];
                    aliases.insert(
                        (block.id.clone(), format!("out{port}")),
                        if conditional_outputs.contains(&child.id) {
                            Port {
                                block: child.id.clone(),
                                port: "out".into(),
                            }
                        } else {
                            drivers[&(child.id.clone(), "in".into())].clone()
                        },
                    );
                }
            }
            _ => {}
        }
    }
    aliases
}

fn resolve_alias(
    source: &Port,
    aliases: &BTreeMap<Key, Port>,
    cache: &mut BTreeMap<Key, Port>,
) -> Result<Port, ModelError> {
    let mut p = source.clone();
    let mut visited = BTreeSet::new();
    loop {
        let k = key(&p);
        if let Some(found) = cache.get(&k) {
            p = found.clone();
            break;
        }
        if let Some(next) = aliases.get(&k) {
            if !visited.insert(k) {
                return Err(fail(
                    &p.block,
                    "instantaneous cycle through subsystem boundary ports",
                ));
            }
            p = next.clone();
        } else {
            break;
        }
    }
    for k in visited {
        cache.insert(k, p.clone());
    }
    Ok(p)
}

type Blocks<'a> = BTreeMap<&'a str, &'a Block>;
type Boundaries<'a> = BTreeMap<(Option<&'a str>, bool), BTreeMap<usize, &'a Block>>;

fn validate_parent(block: &Block, blocks: &Blocks<'_>) -> Result<(), ModelError> {
    let mut ancestor = block.parent.as_deref();
    let mut seen = BTreeSet::from([block.id.as_str()]);
    while let Some(id) = ancestor {
        let parent = blocks
            .get(id)
            .ok_or_else(|| fail(&block.id, "parent system is missing"))?;
        if !matches!(parent.kind, BlockKind::Subsystem { .. })
            || !seen.insert(id)
            || seen.len() > 33
        {
            return Err(fail(
                &block.id,
                "parent must be a subsystem without cycles, at most 32 levels deep",
            ));
        }
        ancestor = parent.parent.as_deref();
    }
    Ok(())
}

fn validate_hierarchy<'a>(
    model: &'a Model,
    blocks: &Blocks<'a>,
) -> Result<Boundaries<'a>, ModelError> {
    if let Some(id) = model
        .sample_times
        .keys()
        .find(|id| !blocks.contains_key(id.as_str()))
    {
        return Err(
            ModelError::new("sample_time", "sampleTimes references an unknown block").at(id, None),
        );
    }
    let mut boundaries = Boundaries::new();
    for block in &model.blocks {
        validate_parent(block, blocks)?;
        let boundary = match block.kind {
            BlockKind::Inport { port, .. } => Some((true, port)),
            BlockKind::Outport { port } => Some((false, port)),
            _ => None,
        };
        if let Some((input, port)) = boundary
            && boundaries
                .entry((block.parent.as_deref(), input))
                .or_default()
                .insert(port, block)
                .is_some()
        {
            return Err(fail(&block.id, "boundary port number is duplicated"));
        }
        if let BlockKind::Inport { data, .. } = &block.kind {
            if block.parent.is_some() && data.is_some() {
                return Err(fail(
                    &block.id,
                    "internal Inport takes its signal from the parent boundary, not embedded data",
                ));
            }
            if block.parent.is_none() && data.is_none() {
                return Err(ModelError::new(
                    "input_data",
                    "root Inport needs an embedded time series",
                )
                .at(&block.id, None));
            }
        }
        if model.sample_times.contains_key(&block.id)
            && (matches!(block.kind, BlockKind::Subsystem { .. })
                || boundary.is_some() && block.parent.is_some())
        {
            return Err(fail(
                &block.id,
                "virtual subsystem and internal port rates are inherited; configure a numerical block instead",
            ));
        }
    }
    for ((parent, input), ports) in &boundaries {
        let expected = match parent.and_then(|id| blocks.get(id)).map(|b| &b.kind) {
            Some(BlockKind::Subsystem {
                inputs, outputs, ..
            }) => {
                if *input {
                    *inputs
                } else {
                    *outputs
                }
            }
            _ => ports.len(),
        };
        if ports.len() != expected || !ports.keys().copied().eq(1..=expected) {
            return Err(fail(
                parent.unwrap_or("root"),
                "boundary ports must be unique and contiguous and match the subsystem interface",
            ));
        }
    }
    for b in &model.blocks {
        if let BlockKind::Subsystem {
            inputs, outputs, ..
        } = b.kind
        {
            for (input, count) in [(true, inputs), (false, outputs)] {
                if boundaries
                    .get(&(Some(b.id.as_str()), input))
                    .map_or(0, BTreeMap::len)
                    != count
                {
                    return Err(fail(
                        &b.id,
                        "subsystem is missing an internal boundary port",
                    ));
                }
            }
        }
    }
    Ok(boundaries)
}

fn validate_connections(
    model: &Model,
    blocks: &Blocks<'_>,
) -> Result<BTreeMap<Key, Port>, ModelError> {
    let definitions: BTreeMap<_, _> = model
        .components
        .iter()
        .map(|d| (d.id.as_str(), d))
        .collect();
    let ports = |b: &Block, output: bool| {
        if let BlockKind::Component { component, .. } = &b.kind {
            return definitions
                .get(component.as_str())
                .map_or_else(Vec::new, |d| {
                    if output { &d.outputs } else { &d.inputs }
                        .iter()
                        .map(|p| p.name.clone())
                        .collect()
                });
        }
        if output {
            super::output_names(&b.kind)
        } else {
            b.kind.inputs()
        }
    };
    let mut drivers = BTreeMap::new();
    for edge in &model.connections {
        let source = blocks
            .get(edge.from.block.as_str())
            .ok_or_else(|| fail(&edge.from.block, "connection source is missing"))?;
        let target = blocks
            .get(edge.to.block.as_str())
            .ok_or_else(|| fail(&edge.to.block, "connection target is missing"))?;
        if source.parent != target.parent {
            return Err(fail(
                &target.id,
                "connections cannot cross system boundaries; use Inport/Outport",
            ));
        }
        if !ports(source, true).contains(&edge.from.port)
            || !ports(target, false).contains(&edge.to.port)
        {
            return Err(fail(&target.id, "connection references an unknown port"));
        }
        if drivers.insert(key(&edge.to), edge.from.clone()).is_some() {
            return Err(fail(&target.id, "input has multiple drivers"));
        }
    }
    for block in &model.blocks {
        for port in ports(block, false) {
            if !drivers.contains_key(&(block.id.clone(), port.clone())) {
                return Err(
                    ModelError::new("missing_input", "required input is not connected")
                        .at(&block.id, Some(&port)),
                );
            }
        }
    }
    Ok(drivers)
}
